FetchMigration/python/orchestrator.py

```diff
@@ -0,0 +1,60 @@
+import argparse
+import logging
+import os
+import subprocess
+
+import migration_monitor
+import pre_migration
+from migration_monitor_params import MigrationMonitorParams
+from pre_migration_params import PreMigrationParams
+
+
+__DP_EXECUTABLE_SUFFIX = "/bin/data-prepper"
+__PIPELINE_OUTPUT_FILE_SUFFIX = "/pipelines/pipeline.yaml"
+
+
+def run(dp_base_path: str, dp_config_file: str, dp_endpoint: str):
+    dp_exec_path = dp_base_path + __DP_EXECUTABLE_SUFFIX
+    output_file = dp_base_path + __PIPELINE_OUTPUT_FILE_SUFFIX
+    pre_migration_params = PreMigrationParams(dp_config_file, output_file, report=True)
+    logging.info("Running pre-migration steps...\n")
+    pre_migration_result = pre_migration.run(pre_migration_params)
+    if len(pre_migration_result.created_indices) > 0:
+        # Kick off a subprocess for Data Prepper
+        logging.info("Running Data Prepper...\n")
+        proc = subprocess.Popen(dp_exec_path)
+        # Data Prepper started successfully, run the migration monitor
+        migration_monitor_params = MigrationMonitorParams(pre_migration_result.target_doc_count, dp_endpoint)
+        logging.info("Starting migration monitor...\n")
+        migration_monitor.run(migration_monitor_params)
+        # Migration ended, the following is a workaround for
+        # https://github.com/opensearch-project/data-prepper/issues/3141
+        if proc.returncode is None:
+            proc.terminate()
```

Comment:
What signal does this send?  I'm hoping that its TERM.  If it doesn't respond to TERM w/in N seconds (maybe 10 or 30), it seems like sending a KILL (-9) is appropriate - though you might want to check that it didn't find more data in the interim and has begun to migrate it, if that's a possibility.  
Generally, this is going to be a difficult concurrency problem to solve and should be solved mostly in DataPrepper (you send a signal, it cleans up and handles its own health, data and consistency)

Reply:
> What signal does this send? I'm hoping that its TERM

`terminate` [sends SIGTERM](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.terminate)

> .... and should be solved mostly in DataPrepper (you send a signal, it cleans up and handles its own health, data and consistency)

Data Prepper handles this already.

> If it doesn't respond to TERM w/in N seconds (maybe 10 or 30), it seems like sending a KILL (-9) is appropriate - though you might want to check that it didn't find more data in the interim and has begun to migrate it, if that's a possibility.

Added an item to the backlog to track this improvement - https://opensearch.atlassian.net/browse/MIGRATIONS-1331
Finding more data is not a concern since the source plugin will use scroll / PIT queries against the source